pkg/autoscaler/autoscaler/autoscaler_manager.go

```diff
@@ -216,3 +233,94 @@ func (am *autoScalerManager) updateTidbClusterAutoScaler(tac *v1alpha1.TidbClust
 	}
 	return err
 }
+
+func (am *autoScalerManager) syncMonitor(tc *v1alpha1.TidbCluster, tac *v1alpha1.TidbClusterAutoScaler) error {
+	if monitorRef := tc.Status.Monitor; monitorRef != nil {
+		autoTcList, err := am.getAutoScaledClusters(tac, []v1alpha1.MemberType{v1alpha1.TiDBMemberType, v1alpha1.TiKVMemberType})
+		if err != nil {
+			return err
+		}
+		clusters := sets.String{}
+		clusterMap := make(map[string]*v1alpha1.TidbCluster)
+		for _, autoTc := range autoTcList {
+			fullName := fmt.Sprintf("%s/%s", autoTc.Namespace, autoTc.Name)
+			clusters.Insert(fullName)
+			clusterMap[fullName] = autoTc
+		}
+
+		monitor, err := am.tmLister.TidbMonitors(monitorRef.Namespace).Get(monitorRef.Name)
+		if err != nil {
+			return err
+		}
+
+		monitoredClusters := sets.String{}
+		newClusterRefs := make([]v1alpha1.TidbClusterRef, 0)
+		for _, tcRef := range monitor.Spec.Clusters {
+			ns := tcRef.Namespace
+			if len(ns) == 0 {
+				ns = monitor.Namespace
+			}
+
+			fullName := fmt.Sprintf("%s/%s", ns, tcRef.Name)
+			monitoredClusters.Insert(fullName)
+			tc, err := am.tcLister.TidbClusters(ns).Get(tcRef.Name)
+			if err != nil {
+				if errors.IsNotFound(err) {
+					continue
+				} else {
+					return err
+				}
+			}
+			tacName, ok := tc.Labels[label.AutoInstanceLabelKey]
+			if !ok || tacName != tac.Name {
+				// Not controlled by this AutoScaler, keep it
+				newClusterRefs = append(newClusterRefs, tcRef)
+				continue
+			}
+		}
+
+		updatedTm := monitor.DeepCopy()
+		toAdd := clusters.Difference(monitoredClusters)
+		for name := range toAdd {
+			cluster := clusterMap[name]
+			newClusterRefs = append(newClusterRefs,
+				v1alpha1.TidbClusterRef{Namespace: cluster.Namespace, Name: cluster.Name})
+		}
+		updatedTm.Spec.Clusters = newClusterRefs
+
+		err = am.updateTidbMonitor(updatedTm)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (am *autoScalerManager) updateTidbMonitor(tm *v1alpha1.TidbMonitor) error {
+	ns := tm.GetNamespace()
+	tmName := tm.GetName()
+	monitorSpec := tm.Spec.DeepCopy()
+
+	// don't wait due to limited number of clients, but backoff after the default number of steps
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		var updateErr error
+		_, updateErr = am.cli.PingcapV1alpha1().TidbMonitors(ns).Update(tm)
+		if updateErr == nil {
+			klog.Infof("TidbMonitor: [%s/%s] updated successfully", ns, tmName)
+			return nil
+		}
+		klog.V(4).Infof("failed to update TidbMonitor: [%s/%s], error: %v", ns, tmName, updateErr)
+		if updated, err := am.tmLister.TidbMonitors(ns).Get(tmName); err == nil {
+			// make a copy so we don't mutate the shared cache
+			tm = updated.DeepCopy()
+			tm.Spec = *monitorSpec
```

Comment:
We should not update the whole spec here, users may manually update the spec or update the `spec.clusters`, we have to append our own TCs only here.
You may also use the patch method used here https://github.com/pingcap/tidb-operator/blob/master/pkg/monitor/monitor/monitor_manager.go#L366

Reply:
Is it ok that we keep appending clusters but we do not remove them?